core/src/main/java/com/scalar/db/storage/cosmos/FilterableScannerImpl.java

```diff
@@ -0,0 +1,99 @@
+package com.scalar.db.storage.cosmos;
+
+import com.azure.cosmos.models.FeedResponse;
+import com.scalar.db.api.Result;
+import com.scalar.db.api.Scan;
+import com.scalar.db.api.Scan.Conjunction;
+import com.scalar.db.api.Scanner;
+import com.scalar.db.util.ScalarDbUtils;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+
+@NotThreadSafe
+public final class FilterableScannerImpl extends ScannerImpl implements Scanner {
+
+  private final Set<Conjunction> conjunctions;
+  private int left;
+
+  @SuppressFBWarnings("EI_EXPOSE_REP2")
+  public FilterableScannerImpl(
+      Scan scan, Iterator<FeedResponse<Record>> recordsPages, ResultInterpreter resultInterpreter) {
+    super(recordsPages, resultInterpreter);
+    this.conjunctions = scan.getConjunctions();
+    this.left = scan.getLimit() > 0 ? scan.getLimit() : -1;
+  }
+
+  @Override
+  @Nonnull
+  public Optional<Result> one() {
+    while (currentPageRecords.hasNext() || recordsPages.hasNext()) {
+      // Return the next record of the current page if there is one
+      if (currentPageRecords.hasNext()) {
+        if (left == 0) {
+          return Optional.empty();
+        }
+        Optional<Result> result = getResultWithFiltering(currentPageRecords.next());
+        if (result.isPresent()) {
+          return result;
+        }
+      } else {
+        // Otherwise, advance to the next page
+        currentPageRecords = recordsPages.next().getResults().iterator();
```

Comment:
Good point. Thank you. Yeah, we don't have to call `next()` if there are no results left that can be returned. Moreover, we don't have to check `left` several times in `one()` since decreasing `left`

Reply:
Fixed in new `FilterableScanner` of 6b1e92a.